### DISCUSSION TITLE
Problem writing commands with a terminal emulation to the serial port (COMn)

### DISCUSSION THREAD

**Discussion by Wintespe** [arendst/Tasmota#19418] [Q&A]
I am using the VisualStudioCode terminal emulation with the PlatformIO (core6.1.11) to flash an ESP8266 device (NodeMcu Pro, external powered) with Tasmota. After flashing, I set some parameters of Tasmota via the serial port.

Send **Var10 Kueche** Tasmota echoes:
```
16:52:48.439 CMD: Var10 Kueche 
16:52:48.444 RSL: RESULT = {"Var10":"Kueche "}
```

Send **Var10**
```
16:53:47.781 CMD: Var10
16:53:47.786 RSL: RESULT = {"Var10":"Kueche "}
```
Everything is fine.

Now I try:

Send **Var10 Küche** results in:
```
16:54:47.945 CMD: che
16:54:47.951 RSL: RESULT = {"Command":"Unknown"}
```

Send **Var10**
```
16:56:35.661 CMD: Var10
16:56:35.666 RSL: RESULT = {"Var10":"Kueche "}
```

The part **Var10 Kü** seems to be eaten. The rest - **che**  - is interpreted as a command that fails. Is there a problem with UTF-8 encoding/decoding for characters outside the ASCII space ?
The "ü" character is given as 0xC3 0xBC.

Access the Tasmota console in a browser and set **Var10 Küche** - via WiFi - is **no** problem at all.

Back to VisualStudioCode and type
Send **Var10** results in:
```
17:19:08.109 CMD: Var10
17:19:08.113 RSL: RESULT = {"Var10":"Küche"}
```
Read back is OK.

The serial monitor of the Arduino IDE shows the same behaviour.

To make sure that the terminals are not in trouble, I link TX directly to RX for an additional test.
Send **Var10 Küche** result in **Var10 Küche**. The result is OK.

What is wrong with sending the command **Var10 Küche** from a terminal emulation?

**Comment by sfromis** [comment]
Serial console input is only intended for 7-bit US-ASCII.

**Comment by Wintespe** [comment]
@sfromis Thank you for your prompt reply.
It is a pity that it is not always possible to carry out a setup after flashing via serial. Unfortunately, I could not find any information about this restriction in the documentation.

**Reply by sfromis** [reply]
What I'm doing for automated post-flash config is to only do a rudimentary config via serial, and then do the rest via MQTT (including chars > 127), via a MQTT client detecting an unconfigured device, to publish the relevant commands from my simple config management system.

Other reasons for me to not use serial as a primary way of doing config include that full speed long commands may not work well (at least not without using a parameter for "slow typing"), and the setup process may take a couple of reboots to be able to config all details.

**Comment by Wintespe** [comment]
For this reason, I am working on a tool that will take over this task for me. The tool reads in a text file and then carries out the configuration. The delay times are programmable like   @Wai_t nnn or _@CheckReady Pre:20; Retry:2; Until:600; Post:100

Here is an example (reduced):

```
BatchID>
@; ----------------------------------------------------------------------
@; 
@; This is the initialisation batch for the PID thermostat controller
@; 
@; ----------------------------------------------------------------------
@; If COM must clear the Interface
@CheckReady Pre:20; Retry:2; Until:600; Post:100

@; ----------------------------------------------------------------------
@; Disable MQTT. Enable HTTP API, Web referrer check for HTTP API commands
Backlog SetOption3 0; SetOption128 #:NoResponse
@;  Please wait upto 9 seconds
@CheckReady Pre:2000; Retry:2; Until:6000; Post:1000

@; ----------------------------------------------------------------------
@; GPIO assignment -> GPIO2=Relay_i1 GPIO4=Relay2 GPIO14=DS18x20
GPIO2 256
GPIO4 225
GPIO14 1312

@; ----------------------------------------------------------------------
@; Enable Belkin WeMo emulation for Alexa
Emulation 1

@; ----------------------------------------------------------------------
@; Set Rules, may take a while
{
Rule1
  On Power1#State = 1 DO PidSp %mem1% ENDON 
  On Power1#State = 0 DO PidSp %mem2% ENDON
}
@; Please wait 3 seconds
@Wait 3
{
Rule2 ON system#boot DO Backlog
  PidSp %mem2%;
  PidTi 80;
  PidTd 25;
  PidPb 1.2;
  PidTc 40;
  PidUpdateSecs 1 ENDON
}
@; Please wait 3 seconds
@Wait 3
Rule1 1
Rule1 5
Rule2 1

@; ----------------------------------------------------------------------
@; Restart device with save the configuration to the flash
Restart 1
@; Please wait 3 seconds
@Wait 3

@; ----------------------------------------------------------------------
@; Setup completed. Do not forget to perform a hardware reset by powering the unit off and on.
@;

```

**Reply by sfromis** [reply]
A couple of details.... Using a template is recommended over messing with individual GPIO commands. Your `SetOption128` command is doing nothing. And if you indend multiple commands at the same time, probably put them in a Backlog, unless you have another mechanism for "dripping" them.

And I still prefer event-based flows instead of using "blind" delay intervals :wink:

**Comment by Wintespe** [comment]
Thnx, no offence.

As an example, I have set the GPIO pins in the example to use  **no** template at the moment. . These settings are well documented by Tasmota.

The commands in curly brackets are processed as a compound line. This is just to show how it works when necessary.

Set option 128 is described in the documentation.

Only Wait nn is a fixed delay - CheckReady tests if Tasmota reacts with the parameter Until:xxx.

**Reply by sfromis** [reply]
Yes, SetOption128 is documented, but the **specific** syntax you use is not setting anything, only querying the current  value. The commands relevant to my observation are not within "curly brackets". And Templates is absolutely recommended for a compound config, after you've created it in the web UI - not by looking up numbers yourself.

**Comment by Wintespe** [comment]
The purpose of the tool is not to think about the extent to which there are other possibilities, such as templates. I think it is inappropriate to discuss details at this point, and that was not my aim.  
The point was to find a solution to parse a text file and then communicate with Tasmota.  This should be possible without MQTT, but simply **without any installations** under Windows
 Until now, this has been done via the serial port or the network.
What the user does with it in detail is up to him. Setting parameters is one possibility.
A terminal emulation with logging to a file is also planned. This was the reason for the request for UTF-8 encoding/decoding on the serial port. 
Originally, a simple way to record the time course of sensors ... in Tasmota and save it to a file was planned.

**Comment by Wintespe** [comment]
Thanks for the hint. I have corrected the batch above to
 `Backlog SetOption3 0; SetOption128 0 #:NoResponse`
